shell/utils/resourceWatcher.js

```diff
@@ -0,0 +1,212 @@
+            } = watch;
+            const watchObject = {
+              resourceType,
+              id,
+              namespace,
+              selector
+            };
+            const watchKey = keyForSubscribe(watchObject);
+
+            if (status === PENDING) {
+              this.send(JSON.stringify({
+                ...watchObject,
+                resourceVersion: !skipResourceVersion ? resourceVersion : undefined
+              }));
+              this.watches[watchKey].status = REQUESTED;
+            } else if (watch.status === CANCELLED) {
+              this.send(JSON.stringify({
+                ...watchObject,
+                stop: true
+              }));
+              this.watches[watchKey].status = REMOVED;
+            }
+          });
+      }
+    } else if (watchesArray.length > 0) {
+      watchesArray
+        .forEach((watch) => {
+          const { status } = watch;
+          const watchKey = keyForSubscribe(watch);
+
+          if ([PENDING, REQUESTED, WATCHING, REFRESHING].includes(status)) {
+            this.watch(watchKey);
+          } else {
+            this.watch[watchKey].status = REMOVED;
+          }
+        });
+
+      this.connect(this.connectionMetadata);
+    }
+  }
+
+  cleanupWatches() {
+    // get rid of all of the removed watches
+    Object.values(this.watches)
+      .filter(watch => watch.status === WATCHSTATUSES.REMOVED)
+      .forEach((watch) => {
+        delete this.watches[keyForSubscribe(watch)];
+      });
+  }
+
+  _onmessage(event) {
+    const {
+      REQUESTED, WATCHING, STOPPED, CANCELLED, REMOVED
+    } = WATCHSTATUSES;
+
+    const {
+      name: eventName, resourceType, data: { type }, id, namespace, selector, data
+    } = JSON.parse(event.data);
+    const watchKey = keyForSubscribe({
+      resourceType,
+      type,
+      id,
+      namespace,
+      selector
+    });
+
+    if (eventName === 'resource.start' && this.watches?.[watchKey]?.status === REQUESTED) {
+      this.watches[watchKey].status = WATCHING;
+    } else if (eventName === 'resource.stop' && this.watches?.[watchKey] && ![CANCELLED, REMOVED].includes(this.watches?.[watchKey]?.status)) {
+      this.watches[watchKey].status = STOPPED;
+      delete this.watches[watchKey].resourceVersion;
+      delete this.watches[watchKey].resourceVersionTime;
+      this.watch(watchKey);
+    } else if (eventName === 'resource.error') {
+      const err = data?.error?.toLowerCase();
+
+      if ( this.watches[watchKey] && err.includes('watch not allowed') ) {
+        this.watches[watchKey].error = { type: resourceType, reason: NO_WATCH };
+        this.unwatch(watchKey);
+      } else if ( this.watches[watchKey] && err.includes('failed to find schema') ) {
+        this.watches[watchKey].error = { type: resourceType, reason: NO_SCHEMA };
+        this.unwatch(watchKey);
+      } else if ( err.includes('too old') ) {
+        this.watch(watchKey);
```

Comment:
We still need to the resyncWatch functionality here

Reply:
unless I'm missing something, I think the watch function here should cover everything that needs to happen on "too old" for us, including getting the new revision.

Reply:
`watch` will only fetch a new revision if the resourveVersion states are cleared or invalid. We're going on the assumption it's 5 minutes, depending on server setting could skip if it's not hit.

also `resyncWatch` removes stale entries

Reply:
so this'll now clear out the resourceVersion and resourceVersionTime which should force it to go get a new one.

I also set this to emit a resyncWatch event message after the watch is completed which will be passed from the worker to the UI thread.